### ISSUE TITLE
hasOne() relationship not working properly in 2.0.x

### ISSUE THREAD

**Issue by makerlabs** [phalcon/cphalcon#11726]
Lets say we have something like this:
User model:

```php
use Phalcon\Mvc\Model\Validator\Email as Email;

class User extends \Phalcon\Mvc\Model
{
    public $id;

    public $email;

    public $password;

    public function initialize()
    {
        $this->hasOne('id', 'UserProfile', 'user_id', array('alias' => 'UserProfile'));
    }

    public function getSource()
    {
        return 'user';
    }

}
```

UserProfile model:

```php
class UserProfile extends \Phalcon\Mvc\Model
{
    public $user_id;

    public $fullname;

    public function initialize()
    {
        $this->belongsTo('user_id', 'User', 'id', array('alias' => 'User'));
    }

    public function getSource()
    {
        return 'user_profile';
    }

}
```

And we try to assign two different User object instances:

```php
$profile = new \UserProfile();

/**
 * Our first object without email and password values 
 */
$profile->user = new \User();

/**
 * Our second object with email and password values 
 */
$user = new \User();
$user->email = 'test@test.com';
$user->password = '1234567890';

$profile->user = $user;

/* We check the values just to be sure */
echo "email=".$profile->user->email."<br>";
echo "password=".$profile->user->password."<br>";
 
$profile->fullname = 'Jan Kowalski';

if (!$profile->save()) {
    foreach ($profile->getMessages() as $m) {
        echo $m->getMessage()."<br>";
    }

}
else {
    echo "ok!";
}
```

Our result:
```
email=test@test.com
password=1234567890
email is required
password is required
```

It seems like phalcon is trying to save the first User object rather then the second one.

**Comment by Jurigag** [comment]
```'alias' => 'User'```

But you are using 

```$profile->user```

**Comment by makerlabs** [comment]
@Jurigag I use it properly u use the lowercase lookup the documentation. Actually in this case i could omit the alias because there is no conflict in the relations.

**Comment by Jurigag** [comment]
Well then there must be mistake in documentation. It should be actually used as lowercase.

**Comment by makerlabs** [comment]
@Jurigag ok but this isn't the issue here the issue here is the object assign not the case of the variable...

**Comment by Jurigag** [comment]
Well actually im checking the source and it's always setting to lower so there is no difference.

**Comment by makerlabs** [comment]
Yeah your resolution is working fine which is strange because accodring to the code it should work the same. Maybe it's more like a zephir bug in this case?

**Comment by Jurigag** [comment]
@makerlabs refresh page, i wrote why it's not working :)

**Comment by makerlabs** [comment]
@Jurigag According totally missed that you're right :)

**Comment by Jurigag** [comment]
Well it could be easly fixed in zephir if it could support references(but it doesn't). Right now the fix to this issue(if it's not intentional) would be to set related values actually when saving/creating them, not when setting property.

**Comment by makerlabs** [comment]
@Jurigag I don't use this code this was a mistake I did accidentally in my application and because of the issue that I pointed out here I spend about half an hour to trac it down. So this should be fixed because it's another inconsistency in the framework which leads to strange behaviors don't u agree?

**Comment by Jurigag** [comment]
But it can't be fixed. It just has to be changed so it will don't use __set anymore and it's a huge change in core.

**Comment by makerlabs** [comment]
No it doesn't the problem is here:
https://github.com/phalcon/cphalcon/blob/2.0.x/phalcon/mvc/model.zep#L4096
and here:
https://github.com/phalcon/cphalcon/blob/2.0.x/phalcon/mvc/model.zep#L4232
This is done for optimization but we don't have to do this this way we could check the 
`this->_related[lowerProperty]`
Which is populated anyway in __get and return it's result. The behavior won't change but it will fix the inconsistency. 
What do you think?

**Comment by virgofx** [comment]
This is related to #10800 - We probably need to get input from @andresgutierrez on the rational of why the alias is getting set lowercase.

**Comment by makerlabs** [comment]
Regardless of the argument why do we use the lowercase version in this case this should be change. It's neither logical nor readable solution. I can understand the problem of breaking things that's why we can change this behavior by adding a simple parameter in framework configuration and add a simple upgrade readme. Phalcon gains it's popularity because of the performance by not improving it (not removing this kind of strange issues for sake of compatibility arguments) leads only to make it less popular if you want to go this direction by all means please do.

**Comment by Jurigag** [comment]
You don't understand a problem. If you first set alias, it will be added to object as property which is ACCESSIBLE, then if you will set it again it will just change value of property without changing value in _related array beacause property now is accessible beacause it exists in object and is public.

**Comment by sergeyklay** [comment]
Fixed in the `2.1.x` branch.
